AEP/aep-00004.md

```diff
@@ -0,0 +1,193 @@
+
+## Sharded architecture
+![sharding schema](./images/sharding.png)
+
+What we can see is that a sharded database consists of several replica sets each containing only parts of the data. These replica sets are agnostic of each other. Also, from the database user point of view a sharded database is transparent, as if it was communicating with a simple standalone (or replica set if the user wants to specify a read preferency) since the user only communicates with the router.
+
+Hence, the real database is actually the concatenation of numerous and independant logical databases (three in our example) in the MongoDB cluster. The router instance is reponsible of this concatenation work when the user requests multi-documents read operations. 
+
+The router is also able to perform read preferences when querrying a specific shard. For readability concerns, the arrows representing distributed read operations point at the replica sets but keep in mind that the router is able to query the all the members of a shard individually. 
+
+In a sharded cluster, the mongos instance kind of endorses the role of the database user in the previous architectures shown.
+
+Note that the config server is also deployed as a replica set. It is not clear in the schema above because we didn't want to overload it with too much information, as we are focusing on sharding user-intended database.
+
+Finally, also note that multiple mongos routers can be deployed in the cluster too but no need for replica set as they are stateless.
+
+# Rationale
+Sharding the database would alleviate the tension on all the nodes. For the primary nodes, the write operations would be distributed among other shards' primary node, and for the secondary nodes there can be some situations where only the right shard is queried (it will be explicited below in this AEP) and as mentionned above, the read operations are faster on a sharded cluster since the data is distributed among multiple instances.
+
+This sharded (or not) MongoDB deployment would be defined in example our infrastructure code as a Helm release resource (called by Terraform). A Helm release is an instance of a Helm chart. Helm charts are templates that allow to deploy resources within a Kubernetes cluster in configurable way.
+Using Helm charts conveys the complexity of defining a deployment towards the authors of a Helm chart, in accordance with the "Don't reinvent the wheel" principle.
+
+Instead of  manually defining all the Kubernetes resources and their configuration for a MongoDB deployment within Terraform modules as it is currently done with ArmoniK's example infrastructure, Helm charts allow us to focus only on passing the good configuration for our MongoDB deployment, which is already not a simple task. Nevertheless, it is possible that some Kubernetes resources or bash scripts will still need to be manually defined inside Terraform for configuration purposes. Helm charts can offer that possibility.
```

Comment:
```suggestion
Instead of manually defining all the Kubernetes resources and their configuration for a MongoDB deployment within Terraform modules as it is currently done with ArmoniK's example infrastructure, Helm charts allow us to focus only on passing the desired configuration for our MongoDB deployment. Nevertheless, it is possible that some Kubernetes resources or bash scripts will still be needed inside Terraform for specific configuration purposes. Helm charts is also offering that capacity.
```